Offset metrics now supports self-offset management.

pennello opened the pull request:
[OTPL-2035][1]

[1]: https://opentable.atlassian.net/browse/OTPL-2035

mikebell90 commented on src/main/java/com/opentable/kafka/util/OffsetMetrics.java:
```diff
@@ -197,33 +215,48 @@ private void poll() {
-        final Map<Integer, Long> lag;
+        final boolean expectKafkaManagedOffsets = offsetsSupplier == null;
+        final Supplier<Map<Integer, Long>> offsetsSupplier;
+        if (expectKafkaManagedOffsets) {
```
This can all be built in constructor.

pennello commented:
@mikebell90 Fixed the offsets suppler to take a topic in 355249f57d86887476bd1e1e901688976ae936cd.

Addressed your concerns about the "limited" functions in 355249f57d86887476bd1e1e901688976ae936cd...c1467ded04a2368c0e042cf8b02c088aa0564d7c.

pennello commented:
Thanks Mike!

stevenschlansker commented on src/main/java/com/opentable/kafka/util/OffsetMetrics.java:
```diff
@@ -77,10 +95,12 @@
     private final OffsetMonitor monitor;
+    private final boolean expectKafkaManagedOffsets;
+    /** Topic -> partition -> offset. */
+    private final Function<String, Map<Integer, Long>> offsetsSupplier;
```
FWIW, I chose to assume dense partitions and used `long[]` in other places.

pennello replied:
Cool, good context.

stevenschlansker commented on src/main/java/com/opentable/kafka/util/OffsetMetrics.java:
```diff
@@ -234,21 +269,85 @@ private void pollUnsafe(final String topic) {
+    /**
+     * Sub-map of {@code map} whose keys are the elements of {@code keys} with corresponding values from {@code map}.
+     */
+    private static <K, V> Map<K, V> subMap(final Collection<K> keys, final Map<K, V> map) {
```
Is this equivalent to a construction like
```java
Maps.filterKeys(map, keys::contains)
```

pennello replied:
Yes.  Implemented in b25d714e1da435924a022b8a7b3f25c80ba116b2.

pennello replied:
(Thank you for the suggestion!!)

stevenschlansker commented on src/main/java/com/opentable/kafka/util/OffsetMetrics.java:
```diff
@@ -234,21 +269,85 @@ private void pollUnsafe(final String topic) {
+
+    @VisibleForTesting
+    static class LongGauge implements Gauge<Long>, Counting {
+        private final AtomicReference<Long> value = new AtomicReference<>();
```
You could hold an `AtomicLong` instead of `AtomicReference<Long>` or if you really want to you could even `extends AtomicLong`

pennello replied:
Or moreover, use our `AtomicLongGauge`?

No to all three.  `AtomicLong`s can't be `null`.

pennello commented:
Streamlined sanity checking in c4722c87eba16a7cbb11c52f9efe2a9dafac7cd1.
